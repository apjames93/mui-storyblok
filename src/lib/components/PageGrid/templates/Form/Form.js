/* eslint-disable no-undef */
/* eslint-disable max-len */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Typography } from '@material-ui/core';
import {
  nestedComponentsProps,
  validComponents,
  muiStringProp,
} from 'lib/utils/customProps';
import {
  Form as RffForm,
  MuiInput,
  MuiSelect,
  MuiSubmit,
  MuiCheckbox,
  MuiRadio,
} from 'rff-wrapper';

import Grid from 'lib/components/PageGrid/templates/Grid/Grid';
import GridItem from 'lib/components/PageGrid/templates/GridItem/GridItem';
import GoogleRecaptcha from './components/GoogleRecaptcha/GoogleRecaptcha';

const components = {
  MuiInput,
  MuiSelect,
  MuiSubmit,
  MuiCheckbox,
  MuiRadio,
  FormGrid: Grid,
  FormGridItem: GridItem,
};

const Form = ({
  content,
  submitButton,
  googleRecaptcha,
  baseUrl,
  successResponseText,
  errorResponseText,
  method,
  storyblokClass,
  dataBlokC,
  dataBlokUid,
}) => {
  const [state, setState] = useState({
    successResponse: '',
    errorResponse: '',
    isABot: true,
  });

  const onSubmit = async (values) => {
    setState({
      successResponse: successResponseText,
      errorResponse: '',
    });
    const queryString = Object.keys(values).map(key => `${key}=${values[key]}`).join('&');
    const options = {
      method,
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      data: queryString,
    };

    fetch(baseUrl, options)
      .then((response) => {
        if (response.ok) {
          setState({
            successResponse: successResponseText,
            errorResponse: '',
          });
          return state;
        }
      }).catch(() => setState({ successResponse: '', errorResponse: errorResponseText }));
  };

  return (
    <RffForm
      onSubmit={onSubmit}
      className={storyblokClass}
      data-blok-c={dataBlokC}
      data-blok-uid={dataBlokUid}
    >
      <>
        {content.map((item, index) => (
          <Grid {...item} key={index} components={components} />
        ))}
        {googleRecaptcha.length
          ? (
            <>
              <GoogleRecaptcha {...googleRecaptcha[0]} isABot={state.isABot} setIsABot={setState} />
              <div style={state.isABot ? { display: 'none' } : null}>
                <MuiSubmit {...submitButton[0]} />
              </div>
            </>
          )
          : <MuiSubmit {...submitButton[0]} />}
        {state.successResponse && <Typography data-testid="successResponseTestID">{state.successResponse}</Typography>}
        {state.errorResponse && <Typography color="error" data-testid="errorResponseTestID">{state.errorResponse}</Typography>}
      </>
    </RffForm>
  );
};

export default Form;

Form.propTypes = {
  content(props, propName, componentName) {
    const comps = [
      'MuiInput',
      'MuiSelect',
      'MuiCheckbox',
      'MuiRadio',
    ];
    return nestedComponentsProps(props, propName, componentName, comps);
  },
  submitButton(props, propName, componentName) {
    const comps = ['MuiSubmit'];
    return validComponents(props, propName, componentName, comps, 1);
  },
  googleRecaptcha(props, propName, componentName) {
    const comps = ['GoogleRecaptcha'];
    return validComponents(props, propName, componentName, comps, 1);
  },
  method(props, propName, componentName) {
    const validProps = ['POST', 'GET'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  baseUrl: PropTypes.string.isRequired,
  successResponseText: PropTypes.string.isRequired,
  errorResponseText: PropTypes.string.isRequired,
};

Form.defaultProps = {
  method: 'POST',
  content: [],
  submitButton: [],
  googleRecaptcha: [],
};
