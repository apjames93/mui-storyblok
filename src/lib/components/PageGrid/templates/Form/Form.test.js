import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import Form from './Form';

function setup() {
  const props = {
    _editable: '<!--#storyblok#{"name": "Form", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    content: [{
      _editable: '<!--#storyblok#{"name": "FormGrid", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'FormGrid',
      content: [{
        _editable: '<!--#storyblok#{"name": "FormGridItem", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'FormGridItem',
        content: [{
          _editable: '<!--#storyblok#{"name": "MuiCheckbox", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiCheckbox',
          name: 'checkbox',
        }, {
          _editable: '<!--#storyblok#{"name": "MuiInput", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiInput',
          name: 'input',
        }, {
          _editable: '<!--#storyblok#{"name": "MuiSelect", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiSelect',
          name: 'select',
          options: ['one', 'two', 'three'],
        }, {
          _editable: '<!--#storyblok#{"name": "MuiRadio", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiRadio',
          name: 'radio',
          value: 'radio1',
        }, {
          _editable: '<!--#storyblok#{"name": "MuiRadio", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
          component: 'MuiRadio',
          name: 'radio',
          value: 'radio2',
        }],
      }],
    }],
    submitButton: [{
      buttonText: 'submit',
    }],
    baseUrl: 'woo.com',
    successResponseText: '!!!!!!!!!!!!!!!!!!!!!good!!!!!!!!!!!!!!!!!!!!!',
    errorResponseText: 'bad',
  };
  const comp = mount(<Form {...props} />);
  return { comp, props };
}

describe('<Form />', () => {
  it('renders Form', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });


  describe('Fetch Request should set success and error responses based on request correctly', () => {
    let container;

    beforeEach(() => {
      container = document.createElement('div');
      document.body.appendChild(container);
    });

    afterEach(() => {
      document.body.removeChild(container);
      container = null;
    });

    it('clicks submit set successResponseText', async () => {
      global.fetch = jest.fn().mockImplementationOnce(() => Promise.resolve({
        ok: true,
        status: 200,
      }));
      const { props } = setup();
      act(() => {
        ReactDOM.render(<Form {...props} />, container);
      });

      const button = container.querySelector('button');
      await act(async () => {
        // eslint-disable-next-line no-undef
        button.dispatchEvent(new MouseEvent('click'));
      });
      expect(container.innerHTML.includes('data-testid="successResponseTestID"')).toBe(true);
    });

    it('clicks submit set errorResponse', async () => {
      global.fetch = jest.fn(() => Promise.reject());
      const { props } = setup();
      act(() => {
        ReactDOM.render(<Form {...props} />, container);
      });

      const button = container.querySelector('button');
      await act(async () => {
        // eslint-disable-next-line no-undef
        button.dispatchEvent(new MouseEvent('click'));
      });
      expect(container.innerHTML.includes('data-testid="errorResponseTestID"')).toBe(true);
    });
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create((
      <Form {...props} />
    ));
    expect(tree).toMatchSnapshot();
  });
});
