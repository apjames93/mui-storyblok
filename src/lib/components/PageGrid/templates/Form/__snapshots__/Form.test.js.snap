// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiBox-root MuiBox-root-63"
  >
    <div
      className="MuiGrid-root makeStyles-root-62  MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
      data-blok-c=""
      data-blok-uid=""
    >
      <div
        className="MuiGrid-root makeStyles-root-66 FormGridItem storyblok--outline MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
        data-blok-c="{\\"name\\": \\"FormGridItem\\", \\"space\\": \\"48408\\", \\"uid\\": \\"947be9f0-47c3-4315-a95a-550f0c560eb5\\", \\"id\\": \\"307934\\"}"
        data-blok-uid=""
      >
        <label
          className="MuiFormControlLabel-root"
          field={
            Object {
              "input": Object {
                "checked": false,
                "name": "checkbox",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "type": "checkbox",
                "value": null,
              },
              "meta": Object {
                "active": false,
                "data": Object {},
                "dirty": false,
                "dirtySinceLastSubmit": false,
                "error": undefined,
                "initial": false,
                "invalid": false,
                "length": undefined,
                "modified": false,
                "modifiedSinceLastSubmit": false,
                "pristine": true,
                "submitError": undefined,
                "submitFailed": false,
                "submitSucceeded": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "validating": false,
                "visited": false,
              },
            }
          }
          id="checkbox"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-10 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                className="PrivateSwitchBase-input-13"
                data-indeterminate={false}
                disabled={false}
                onChange={[Function]}
                type="checkbox"
                value={null}
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            
          </span>
        </label>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          data-testid="muiinput"
          field={
            Object {
              "input": Object {
                "checked": undefined,
                "name": "input",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "type": "text",
                "value": "",
              },
              "meta": Object {
                "active": false,
                "data": Object {},
                "dirty": false,
                "dirtySinceLastSubmit": false,
                "error": undefined,
                "initial": undefined,
                "invalid": false,
                "length": undefined,
                "modified": false,
                "modifiedSinceLastSubmit": false,
                "pristine": true,
                "submitError": undefined,
                "submitFailed": false,
                "submitSucceeded": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "validating": false,
                "visited": false,
              },
            }
          }
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            input
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete=""
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              name="input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              rows=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
          field={
            Object {
              "input": Object {
                "checked": undefined,
                "name": "select",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "value": "",
              },
              "meta": Object {
                "active": false,
                "data": Object {},
                "dirty": false,
                "dirtySinceLastSubmit": false,
                "error": undefined,
                "initial": undefined,
                "invalid": false,
                "length": undefined,
                "modified": false,
                "modifiedSinceLastSubmit": false,
                "pristine": true,
                "submitError": undefined,
                "submitFailed": false,
                "submitSucceeded": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "validating": false,
                "visited": false,
              },
            }
          }
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            select
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="mui-component-select-select"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="mui-component-select-select"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              autoComplete=""
              autoFocus={false}
              className="MuiSelect-nativeInput"
              name="select"
              onAnimationStart={[Function]}
              onChange={[Function]}
              placeholder=""
              required={false}
              tabIndex={-1}
              value=""
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <label
          className="MuiFormControlLabel-root"
          data-testid="MuiRadioID"
          field={
            Object {
              "input": Object {
                "checked": false,
                "name": "radio",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "type": "radio",
                "value": "radio1",
              },
              "meta": Object {
                "active": false,
                "data": Object {},
                "dirty": false,
                "dirtySinceLastSubmit": false,
                "error": undefined,
                "initial": undefined,
                "invalid": false,
                "length": undefined,
                "modified": false,
                "modifiedSinceLastSubmit": false,
                "pristine": true,
                "submitError": undefined,
                "submitFailed": false,
                "submitSucceeded": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "validating": false,
                "visited": false,
              },
            }
          }
          id="radio"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-10 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-13"
                disabled={false}
                onChange={[Function]}
                type="radio"
                value="radio1"
              />
              <div
                className="PrivateRadioButtonIcon-root-14"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-15"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </div>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            
          </span>
        </label>
        <label
          className="MuiFormControlLabel-root"
          data-testid="MuiRadioID"
          field={
            Object {
              "input": Object {
                "checked": false,
                "name": "radio",
                "onBlur": [Function],
                "onChange": [Function],
                "onFocus": [Function],
                "type": "radio",
                "value": "radio2",
              },
              "meta": Object {
                "active": false,
                "data": Object {},
                "dirty": false,
                "dirtySinceLastSubmit": false,
                "error": undefined,
                "initial": undefined,
                "invalid": false,
                "length": undefined,
                "modified": false,
                "modifiedSinceLastSubmit": false,
                "pristine": true,
                "submitError": undefined,
                "submitFailed": false,
                "submitSucceeded": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "validating": false,
                "visited": false,
              },
            }
          }
          id="radio"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-10 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-13"
                disabled={false}
                onChange={[Function]}
                type="radio"
                value="radio2"
              />
              <div
                className="PrivateRadioButtonIcon-root-14"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-15"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </div>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            
          </span>
        </label>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-fullWidth"
    data-testid="Button"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      submit
    </span>
  </button>
  
  
</form>
`;
